scripts/cameo-tag.py

```diff
@@ -0,0 +1,115 @@
+# We will reuse the last bytes of the IH_NAME field. This is the tricky 
+# part because we have the header CRC and the CAMEO checksum that must match
+# the whole header. uImage header CRC checksums all data expect the CRC data
+# itself. The same applies for CAMEO checksum. It checksums all data except
+# itself. Changing one of both results in a change of the other. So we will
+# brute force checksum combinations until we get an exact match between 
+# those two. This method usually converges after ~500 iterations (<0.2s) and 
+# is easier than using reverse math. After manipulation the uImage header
+# looks like this:
+#
+# [....<header_crc>....<shortened_name><00><CAMEOTAG><0001><checksum>] 
+#
+# Location for original rootfs partition is very simple. It is behind the 
+# OpenWRT compressed kernel image file that spans into the rootfs. So
+# the tag will be written somewhere to the following rootfs partition and 
+# can be found by U-Boot. The CAMEO checksum calculation must start at
+# position 0x180000 (see DLINK_KERNEL_PART_SIZE) and includes the "second"
+# half of the "split" kernel uImage. 
+
+import argparse
+import pathlib
+import socket
+import struct
+import zlib
+
+UIMAGE_HEADER = 64
+CAMEO_TAG = bytearray([0x43, 0x41, 0x4d, 0x45, 0x4f, 0x54, 0x41, 0x47, 0x00, 0x00, 0x00, 0x01])
+NO_CRC = bytearray([0, 0, 0, 0])
+
+def read_buffer(offset, whence, count):
+    args.uimage_file.seek(offset, whence)
+    return bytearray(args.uimage_file.read(count))
+
+def write_buffer(offset, whence, buf):
+    args.uimage_file.seek(offset, whence)
+    args.uimage_file.write(buf)
+
+def toarray(value):
+    return value.to_bytes(4, 'big')
+
+def cameosum(array):
+    return sum(iter(array)) % (1<<32)
+
+def guess_crc_checksum(buf):
+    buf[46] = 0 # cut image name after 14 chars
+    buf[48:60] = CAMEO_TAG
+
+    for i in range(20):
+        buf[47] = i # overflow counter for additional checksum combinations
+        buf[4:8] = NO_CRC
+        startsum = cameosum(buf[0:60])
+        crc = zlib.crc32(buf[0:60])
+        for testsum in range(startsum, startsum+255*4):
+            if testsum == startsum + cameosum(toarray(zlib.crc32(toarray(testsum), crc))):
+                buf[4:8] = toarray(zlib.crc32(toarray(testsum), crc))
+                buf[60:64] = toarray(testsum)
+                return buf
+    return None
+
+parser = argparse.ArgumentParser(description='Insert Cameo firmware tags.')
+parser.add_argument('uimage_file', type=argparse.FileType('r+b'))
+parser.add_argument('rootfs_start', type=int)
+args = parser.parse_args()
+
+# tag for the second part of the kernel that resides in the
+# vendor rootfs partition. For this we will add the CAMEO tag
+# and the checksum to the end of the image.
+
+buf = read_buffer(args.rootfs_start, 0, -1)
+if len(buf) == 0:
+    raise ValueError(f"uImage smaller than " + str(args.rootfs_start) + " bytes")
+
+write_buffer(0, 2, CAMEO_TAG + toarray(cameosum(buf + CAMEO_TAG)))
```

Comment:
Can you `import os` so you can use `os.SEEK_END` etc. for whence? I don't like seeing magic constants.